processing/src/main/java/io/druid/query/groupby/epinephelinae/StreamingMergeSortedGrouper.java

```diff
@@ -0,0 +1,387 @@
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int notUsed)
+  {
+    return aggregate(key);
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    try {
+      final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+
+      if (keyBuffer.remaining() != keySize) {
+        throw new IAE(
+            "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+            keyBuffer.remaining(),
+            keySize
+        );
+      }
+
+      final int prevRecordOffset = curWriteIndex * recordSize;
+      if (curWriteIndex == -1 || !keyEquals(keyBuffer, buffer, prevRecordOffset)) {
+        initNewSlot(keyBuffer);
+      }
+
+      final int curRecordOffset = curWriteIndex * recordSize;
+      for (int i = 0; i < aggregatorOffsets.length; i++) {
+        aggregators[i].aggregate(buffer, curRecordOffset + aggregatorOffsets[i]);
+      }
+
+      return AggregateResult.ok();
+    }
+    catch (RuntimeException e) {
+      finished = true;
+      throw e;
+    }
+  }
+
+  private boolean keyEquals(ByteBuffer curKeyBuffer, ByteBuffer buffer, int bufferOffset)
+  {
+    int i = 0;
+    for (; i + Long.BYTES <= keySize; i += Long.BYTES) {
+      if (curKeyBuffer.getLong(i) != buffer.getLong(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    if (i + Integer.BYTES <= keySize) {
+      if (curKeyBuffer.getInt(i) != buffer.getInt(bufferOffset + i)) {
+        return false;
+      }
+      i += Integer.BYTES;
+    }
+
+    for (; i < keySize; i++) {
+      if (curKeyBuffer.get(i) != buffer.get(bufferOffset + i)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private void initNewSlot(ByteBuffer newKey)
+  {
+    increaseWriteIndex();
+
+    final int recordOffset = recordSize * curWriteIndex;
+    buffer.position(recordOffset);
+    buffer.put(newKey);
+
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].init(buffer, recordOffset + aggregatorOffsets[i]);
+    }
+  }
+
+  /**
+   * Wait for {@link #nextReadIndex} to be moved if necessary and move {@link #curWriteIndex}.  {@link #nextReadIndex}
+   * is checked in while loops instead of waiting using a lock to avoid frequent thread park.
+   */
+  private void increaseWriteIndex()
+  {
+    if (curWriteIndex == maxSlotNum - 1) {
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextReadIndex == -1 || nextReadIndex == 0) && !Thread.interrupted()) {
+        if (System.currentTimeMillis() - startLoopAt > timeoutMs) {
+          throw new RuntimeException(new TimeoutException());
+        }
+      }
+      curWriteIndex = 0;
+    } else {
+      final int nextWriteIndex = curWriteIndex + 1;
+      final long startLoopAt = System.currentTimeMillis();
+      while ((nextWriteIndex == nextReadIndex) && !Thread.interrupted()) {
```

Comment:
Also, this concurrent code should have a lot of comments, along with every line that has any concurrent meaning, with references like `(*)` or `(1)` between lines, explaining happens-before edges, etc.

E. g. a comment to this line (and similar in spin loops below) should refer to https://stackoverflow.com/questions/11761552

Reply:
Will add.